---
title: "Correcting machine learning models"
output:
  pdf_document:
    keep_tex: true
header-includes:
  \usepackage{caption}
  \usepackage{float}
  \usepackage{amsmath}
  \usepackage{graphicx}
  \floatplacement{figure}{H}
  \captionsetup[figure]{labelfont={bf}, name={Figure}, labelsep=period}
---


```{r, analysis, echo=FALSE}
source("config.R")
source("cml-mnist.R")
source("cml-testis.R")
mnist_example <- 3
```


```{r, graphics.details, echo=FALSE}
knitr::opts_chunk$set(dpi=600, dev="png", dev.args=list(type="cairo-png"))
```


```{r fig.intro, eval=TRUE, echo=FALSE, fig.width=5.3, fig.align="center", fig.cap="\\textbf{Strategies for correcting machine learning models.}  (a) Schematic of a typical machine-learning study with a training dataset and training labels, which can be used to construct a model. The model can be used in production on new input data. (b) Schematic of changes that may require adjusting or correcting an already-trained model. (c) Replacing the original model by a new model constitutes a global strategy for adjusting an existing workflow. (d) Other global strategies for adjusting an existing workflow through pre-processing, post-processing, and use of multiple models in parallel. (e,f) Analogous to (c,d), but showing targeted strategies that affect small portions of the workflow components."}
include_graphics("correcting.png")
```


```{r fig.mnist, eval=TRUE, echo=FALSE, fig.width=6.5, fig.height=6.7, fig.cap="\\textbf{Correcting multi-class classifiers}. \\textbf{(a)} Schematic of partitioning a dataset into sections for trainining, holdout, and testing. The original dataset is used to create variants, which are split into primary and secondary datasets. \\textbf{(b)} Schematic of variants of the MNIST digits dataset; images show how a representative data item is mutated in each variant. \\textbf{(c)} Classification performance obtained by model trained on an MNIST variant with missing pixels. Performance is evaluated on the training images (R), test images with missing pixels (T), and test images from the original dataset (O). \\textbf{(d)} Extension of (c) to all dataset variants. \\textbf{e} Alternative summary of models trained on MNIST variants and evaluated on test images of the original dataset. \\textbf{(f)} Classification performance of a model and an ensemble on test images with increasing amounts of missing pixels. Performance is evaluated with and without a pre-processing filter that imputes missing values by averaging neighboring pixels. \\textbf{(g)} Classification performance of a model and an ensemble on test images with increasing amounts of background noise. Performance is evaluated using models calibrated on noise-free data, and with separate calibration at each noise level."}
m1 <- matrix(c(1,1,1,
               2,3,3,
               4,5,5,
               6,7,7,
               6,7,7,
               8,9,9,
               10,11,12),
             byrow=TRUE, nrow=7, ncol=3)
m2 <- 12 + matrix(c(rep(1, 5), rep(14, 6),
                    rep(1, 5), rep(14, 6),
                    rep(1, 5), rep(14, 6),
                    rep(2, 11),
                    3:13,
                    3:13,
                    3:13),
                  byrow=TRUE, nrow=7, ncol=11)
m3 <- max(m2) + matrix(c(rep(1, 15),
                         rep(2, 4), rep(3, 11)),
                       byrow=TRUE, nrow=2)
layout(rbind(cbind(m1, m2[,1], m2), m3),
       widths=c(0.5, 1.2, 1.0, 0.1, 0.32+0.2, rep(0.32, 10)),
       height=c(1.1, 0.7, 0.7, 0.3, 0.4, 0.7, 0.7, 0.2, 1.9))
mnist_rect <- c(0, 0.1, 0.8, 0.9)
mnist_shift <- c(1, 0, 1, 0)
#
# header row
plot_schematic()
multipanelLabel(panel_labels[1])
#
# new instances
plot_label("change in\ninstances")
multipanelLabel(panel_labels[2])
parplot(0, xlim=c(0, 4), ylim=c(0, 1), Rcssclass="mnist")
add_mnist(mnist$original[mnist_example, ], mnist_rect+mnist_shift*0,
          label="subset")
text(1.0, 0.8, "\U00BD of\ninstances", Rcssclass=c("mnist", "comment"))
# masking features
plot_label("change in\nfeatures")
parplot(0, xlim=c(0, 4), ylim=c(0, 1), Rcssclass="mnist")
add_mnist(mnist$left[mnist_example, ], mnist_rect,
          label="horizontal")
add_mnist(mnist$top[mnist_example, ], mnist_rect+mnist_shift*1,
          label="vertical")
rect(2.0, 0.5, 2.15, 0.7, border=NA,
     col=RcssValue("mnist", "masked", default="#000000"))
text(2.2, 0.6, "pixels reset to 'NA'", Rcssclass=c("mnist", "comment"))
#
# updated content
plot_label("change in\ncontent")
parplot(0, xlim=c(0, 4), ylim=c(0, 1), Rcssclass="mnist")
add_mnist(mnist$dropout[mnist_example, ], mnist_rect,
          label="dropout")
add_mnist(mnist$noise[mnist_example, ], mnist_rect+mnist_shift,
          label="noise")
add_mnist(mnist$erased[mnist_example, ], mnist_rect+mnist_shift*2,
          label="erased")
add_mnist(mnist$spot[mnist_example, ], mnist_rect+mnist_shift*3,
          label="spot")
#
# missing content content
plot_label("change in\ncontent")
parplot(0, xlim=c(0, 4), ylim=c(0, 1), Rcssclass="mnist")
add_mnist(mnist$missing[mnist_example, ], mnist_rect+mnist_shift*0,
          label="missing")
add_mnist(mnist$lost[mnist_example, ], mnist_rect+mnist_shift*1,
          label="lost")
#
# updated labels
plot_label("change in\nlabels")
parplot(0, xlim=c(0, 2), ylim=c(0, 1), Rcssclass="mnist")
add_mnist(mnist$mislabel[mnist_example, ], mnist_rect+mnist_shift*0,
          label="mislabel")
text(1.0, 0.8, "\U00BD of\ninstances\nmislabeled", Rcssclass=c("mnist", "comment"))
ramp_prob <- setNames(rev(1:10), 0:9)
plot_minimal_bars(ramp_prob, ylab="", xlab="", main="ramp")
#
# construct performance object for plotting
perf <- copy(mnist_performance$ensemble)
perf$variant <- gsub("left", "horizontal",
                     gsub("top", "vertical", perf$variant))
perf$css <- perf$label <- perf$rep
perf[label=="1", "label"] <- "model 1"
perf[label=="2", "label"] <- "model 2"
perf[label=="ensemble_uncalibrated", "label"] <- "ensemble\n(uncalibrated)"
perf[label=="ensemble_calibrated", "label"] <- "ensemble\n(calibrated)"
perf[css=="1", "css"] <- "model"
perf[css=="2", "css"] <- "model"
perf$train <- perf$train * 100
perf$test <- perf$test * 100
perf$original_test <- perf$original_test * 100
#
# plot one panel explaining the V-shaped performance plot
perf_zoom <- perf[variant=="missing"]
perf_zoom$label_y <- 0
perf_zoom[rep=="1", "label_y"] <- -2
perf_zoom[rep=="2", "label_y"] <- -5
perf_zoom[rep=="ensemble_calibrated", "label_y"] <- 5
perf_zoom[rep=="ensemble_uncalibrated", "label_y"] <- 1
plot_performance(perf_zoom,
                 states=c(train="training (R)",
                          test="test (T)",
                          original_test="test, original (O)"),
                 ylab="Hit rate (%)",
                 main="Performance of models\n trained w. missing values",
                 ylim=c(84, 100), Rcssclass="zoom")
text(4.5, 81, "evaluation\ndatasets", Rcssclass=c("perf", "zoom"))
multipanelLabel(panel_labels[3])
#
# plot many small panels with V-shaped performance plots
mnist_variants <- c("subset", "horizontal", "vertical",
                    "dropout", "noise", "erased", "spot", "missing", "lost",
                    "mislabel", "ramp")
parplot(c(0, 1), c(0, 1), Rcssclass="perfspace")
text(0.5, 0.3, "Performance of models trained on dataset variants",
     Rcssclass=c("perf", "main"))
multipanelLabel(panel_labels[4], y=1.0)
for (v in mnist_variants) {
  plot_performance(perf[variant==v],
                   states=c(train="R", test="T", original_test="O"),
                   ylab="Hit rate (%)", ylim=c(50, 100),
                   show_y=(v=="subset"), show_label=FALSE,
                   main=gsub("horizontal", "horiz.",
                             gsub("vertical", "vert.", v)),
                   Rcssclass=ifelse(v=="subset", "left", "mid"))
}
#
# plot a single summary
plot_slide(perf, states=mnist_variants,
           state_col="variant", value_col="original_test",
           xlim=c(85, 100), main="Performance on original data")
add_slide_legend <- function(y=3, y_offset=0.5) {
  add_slide_legend_point(c(81, y), c(82, y), "single\nmodel",
                         Rcssclass="model")
  add_slide_legend_point(c(87, y+y_offset), c(88, y+y_offset),
                         "ensemble (uncalibrated)",
                         Rcssclass="ensemble_uncalibrated")
  add_slide_legend_point(c(87, y-y_offset), c(88, y-y_offset),
                         "ensemble (calibrated)",
                         Rcssclass="ensemble_calibrated")
}
add_slide_legend(y=1.3)
multipanelLabel(panel_labels[5])
#
# horizontal filler
par(mai=rep(0.01, 4)); plot.new()
#
# panels with dataset series
imputation <- copy(mnist_performance$imputation)
imputation$series <- paste0(imputation$variant, "_", imputation$rep)
imputation$series <- gsub("ensemble_", "", imputation$series)
imputation$series <- gsub("calibrated", "", imputation$series)
imputation$series <- gsub("_$", "", imputation$series)
imputation$hit_rate <- 100*imputation$test
plot_series(imputation, series_col="series", show_points=FALSE,
            series=c("original", "original_imputed", "subset", "subset_imputed"),
            x_var="na", y_var="hit_rate", xlim=c(0, 80), ylim=c(0, 100),
            xlab="Missingness in test data (%)", ylab="Hit rate (%)",
            main="Correcting for missing data",
            Rcssclass="series80")
add_axis_mnist <- function(x, y=-20, width=5, height=11, prefix="bg") {
  for (.x in x) {
    add_mnist(mnist[[paste0(prefix, "_", .x)]][mnist_example, ],
              c(.x-width/2, y-height/2, .x+width/2, y+height/2))
  }
}
add_axis_mnist(x=c(0, 20, 40, 60, 80), prefix="na")
multipanelLabel(panel_labels[6])
add_series_legend(2, 55, main="Training data", type="rect",
                  labels=c("original (one model)", "subsets (ensemble)"),
                  css=c("original_imputed", "subset_imputed"))
add_series_legend(2, 23, main="Imputation",
                  labels=c("none", "neighbor average"),
                  css=c("neutral", "neutral_imputed"))
#
#
balance <- copy(mnist_performance$calibration)
balance$series <- paste0(balance$variant, "_", balance$rep)
balance$series <- gsub("ensemble_", "", balance$series)
balance$series <- gsub("calibrated", "", balance$series)
balance$series <- gsub("_$", "", balance$series)
balance$hit_rate <- 100*balance$test
plot_series(balance, series_col="series", show_points=FALSE,
            series=c("original", "original_re", "subset", "subset_re"),
            x_var="bg", y_var="hit_rate", xlim=c(0, 100), ylim=c(0, 100),
            xlab="Noise in test data (%)", ylab="Hit rate (%)",
            main="Correcting for background noise")
multipanelLabel(panel_labels[7])
add_axis_mnist(x=c(0, 20, 40, 60, 80, 100), prefix="bg")
add_series_legend(89, 97, main="Training data", type="rect",
                  labels=c("original (one model)", "subsets (ensemble)"),
                  css=c("original_re", "subset_re"))
add_series_legend(89, 65, main="Calibration",
                  labels=c("original data", "noisy data"),
                  css=c("neutral", "neutral_re"))
```


```{r fig.testis, eval=TRUE, echo=FALSE, fig.width=6.5, fig.height=3.2, fig.cap="\\textbf{Correcting cell-type classification}. \\textbf{(a)} Embeddings of gene-expression profiles from single cells from testis. Dots represent individual cells. Embeddings were generated using gene-expression values for 103 marker genes. Panels highlight selected cell types in color; cell types are determined by the dominant gene expression signature. \\textbf{(b)} Performance of cell-type prediction in an independent dataset. Performance was measured for a signature-based classifier without calibration, with calibration to the entire new dataset, and with recalibration to sample-specific data. \\textbf{(c)} Analogous to (b), but with predictions performed using an xgboost multiclass classifier and an ensemble of the xgboost and the signature-based classifiers."}
layout(matrix(c(1, 1, 1, 6, 7,
                2, 3, 4, 6, 8,
                2, 3, 4, 6, 9,
                5, 5, 5, 6, 9),
              byrow=TRUE, ncol=5, nrow=4),
       widths=c(0.9, 0.9, 0.9, 0.2, 3.6),
       height=c(0.2, 1.5, 0.5, 1.0))
#
# panels with unsupervised learning
plot_label("Sohni et al.", y=0.8)
plot_testis_umap <- function(sig, main="") {
  sig_data <- copy(testis_umap)
  sig_data$label <- ifelse(sig_data$top_sig %in% sig, "hit", "aaa")
  plot_umap(sig_data, main=main, point_class_col="label",
            xlim=c(-7.2, 4), ylim=c(-12.5, 16),
            axis_legend_size=c(0.4, 0.22))
  invisible(sig_data)
}
plot_testis_umap("Endothelial", main="Endothelial")
multipanelLabel(panel_labels[1])
plot_testis_umap(c("LC", "PTM"), main="Leydig &\nPeritubular myoid")
plot_testis_umap("SC", main="Sertoli")
#
# horizontal filler and vertical spacer
par(); plot.new();
par(); plot.new();
#
# panels with predictions of cell types
plot_label("Guo et al.", y=0.8)
testis_age_long <- melt(testis_age, id.vars=c("variant", "rep"),
                        measure.vars=grep("test_", colnames(testis_age), value=TRUE),
                        variable.name="age", value.name="test")
testis_age_long$age <- paste0(gsub("test_age_", "", testis_age_long$age), "_yrs")
testis_age_long$test <- 100*testis_age_long$test
testis_age_long$css <- paste0(testis_age_long$variant, "_", testis_age_long$rep)
testis_samples <- c("7_yrs", "11_yrs", "13_yrs", "14_yrs")
plot_slide(testis_age_long[variant=="fun"],
           states=testis_samples,
           state_col="age", value_col="test",
           xlim=c(30, 100), main="Cell type prediction by a signature-based classifier",
           Rcssclass="testis")
multipanelLabel(panel_labels[2])
add_testis_legend <- function(y=1.0, title="Model", x=c(32, 41, 61, 81),
                              x_offset=2, css_prefix="fun") {
  text(x[1], y, title, Rcssclass=c("slide", "testis", "legend"))
  add_slide_legend_point(c(x[2], y), c(x[2]+x_offset, y),
                         "uncalibrated",
                         Rcssclass=c("testis", paste0(css_prefix, "_uncalibrated")))
  add_slide_legend_point(c(x[3], y), c(x[3]+x_offset, y),
                         "calibrated",
                         Rcssclass=c("testis", paste0(css_prefix, "_calibrated")))
  add_slide_legend_point(c(x[4], y), c(x[4]+x_offset, y),
                         "recalibrated",
                         Rcssclass=c("testis", paste0(css_prefix, "_recalibrated")))
}
add_testis_legend(title="Model", x=quantile(c(30, 100), p=c(0, 22, 50, 78)/100),
                    x_offset=(100-30)/40)
plot_slide(testis_age_long[variant %in% c("xgb", "funxgb")],
           states=testis_samples,
           state_col="age", value_col="test",
           xlim=c(94.5, 100), main="Cell type prediction by a dedicated classifier and an ensemble",
           Rcssclass=c("testis"))
multipanelLabel(panel_labels[3])
add_testis_legend(title="Model", y=1.45,
                  x=quantile(c(94.5, 100), p=c(0, 22, 50, 78)/100),
                  x_offset=(100-94.5)/40, css_prefix="xgb")
add_testis_legend(title="Ensemble", y=0.55,
                  x=quantile(c(94.5, 100), p=c(0, 22, 50, 78)/100),
                  x_offset=(100-94.5)/40, css_prefix="funxgb")
```

